synapse/types.py

```diff
@@ -267,6 +267,14 @@ def from_string(cls: Type[DS], s: str) -> DS:
             )
 
         domain = parts[1]
+        if not domain:
+            raise SynapseError(
+                400,
+                f"{cls.__name__} must have a domain after the colon",
+                Codes.INVALID_PARAM,
+            )
+        # TODO: this does not reject an empty localpart or an overly-long string.
+        # See https://spec.matrix.org/v1.2/appendices/#identifier-grammar
```

Comment:
I more checking would get in `is_valid`, not in `from_string`? Since that does additional validation checks.

(I think leaving the domain check here is likely fine, I mean the additional `TODO`.)

Reply:
Depends on your philosophy I suppose; my bias is towards parsers that either output valid data or else complain loudly. But perhaps that is naive.

Reply:
I'll move the comment.

On reflection, it looks like `is_valid` would do what I want it to do, because `parse_server_name` should raise if given an empty `server_name`. I'll remove the extra check from `from_string` and change the endpoint in question to explicitly validate.

Reply:
I *think* the idea is what we call `is_valid` when taking user input and `from_string` when we know it should be valid already?

Reply:
I concur it is all a bit circular...